CollapseLauncher/Classes/Helper/Metadata/DataCooker.cs

```diff
@@ -0,0 +1,160 @@
+
+        internal static string ServeV3Data(string data)
+        {
+            if (!Base64.IsValid(data))
+                return data;
+
+            byte[] dataBytes = Convert.FromBase64String(data);
+            if (IsServeV3Data(dataBytes))
+            {
+                GetServeV3DataSize(dataBytes, out long compressedSize, out long decompressedSize);
+                byte[] outBuffer = new byte[decompressedSize];
+                ServeV3Data(dataBytes, outBuffer, (int)compressedSize, (int)decompressedSize, out int dataWritten);
+                return Encoding.UTF8.GetString(outBuffer.AsSpan(0, dataWritten));
+            }
+
+            return data;
+        }
+
+        internal static bool IsServeV3Data(ReadOnlySpan<byte> data)
+        {
+            if (data.Length < 8) return false;
+
+            long signature = MemoryMarshal.Read<long>(data);
+            if (signature != COLLAPSESIG) return false;
+
+            return true;
+        }
+
+        internal static void GetServeV3DataSize(ReadOnlySpan<byte> data, out long compressedSize, out long decompressedSize)
+        {
+            if (data.Length < 32) throw new FormatException($"The MetadataV3 data format is corrupted!");
+
+            int readOffset = sizeof(long) * 2;
+            compressedSize = MemoryMarshal.Read<long>(data.Slice(readOffset));
+            decompressedSize = MemoryMarshal.Read<long>(data.Slice(readOffset += sizeof(long)));
+        }
+
+        private static void GetServeV3Attribute(ReadOnlySpan<byte> data, out CompressionType compressionType, out bool isUseEncryption)
+        {
+            int readOffset = sizeof(long);
+            long attribNumber = MemoryMarshal.Read<long>(data.Slice(readOffset));
+
+            compressionType = (CompressionType)(byte)attribNumber;
+            isUseEncryption = (byte)(attribNumber >> 8) == 1;
+        }
+
+
+        internal static void ServeV3Data(ReadOnlySpan<byte> data, Span<byte> outData, int compressedSize, int decompressedSize, out int dataWritten)
+        {
+            GetServeV3Attribute(data, out CompressionType compressionType, out bool isUseEncryption);
+            int readOffset = sizeof(long) * 4;
+
+            ReadOnlySpan<byte> dataRawBuffer = data.Slice(readOffset, data.Length - readOffset);
+            byte[] decryptedDataSpan = null;
+            int encBitLength = LauncherMetadataHelper.CurrentMasterKey.BitSize;
+
+            bool isDecryptPoolAllowed = dataRawBuffer.Length <= ALLOWEDBUFFERPOOLSIZE;
+            bool isDecryptPoolUsed = false;
+
+            try
+            {
+                if (isUseEncryption)
+                {
+                    decryptedDataSpan = isDecryptPoolAllowed ? ArrayPool<byte>.Shared.Rent(dataRawBuffer.Length) : new byte[dataRawBuffer.Length];
+                    isDecryptPoolUsed = isDecryptPoolAllowed;
+
+                    if (RSAInstance == null)
+                    {
+                        RSAInstance = RSA.Create();
```

Comment:
The ``RSAInstance`` is initially null. This logic is used for caching to reduce the needs or reinitialization every time the decryption routine is called. Since the ``RSAInstance`` will be used for the rest of the time once it's initialized, we don't need to dispose the instance.